internal/log/log.go

```diff
@@ -0,0 +1,151 @@
+// Package log provides logging utilities for the tracer.
+package log
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/version"
+)
+
+// Level specifies the logging level that the log package prints at.
+type Level int
+
+const (
+	// LevelDebug represents debug level messages.
+	LevelDebug Level = iota
+	// LevelWarn represents warning and errors.
+	LevelWarn
+)
+
+var prefixMsg = fmt.Sprintf("Datadog Tracer %s", version.Tag)
+
+var (
+	mu     sync.RWMutex   // guards below fields
+	level                 = LevelWarn
+	logger ddtrace.Logger = &defaultLogger{l: log.New(os.Stderr, "", log.LstdFlags)}
+)
+
+// UseLogger sets l as the active logger.
+func UseLogger(l ddtrace.Logger) {
+	mu.Lock()
+	defer mu.Unlock()
+	logger = l
+}
+
+// SetLevel sets the given lvl for logging.
+func SetLevel(lvl Level) {
+	mu.Lock()
+	defer mu.Unlock()
+	level = lvl
+}
+
+// Debug prints the given message if the level is LevelDebug.
+func Debug(fmt string, a ...interface{}) {
+	mu.RLock()
+	lvl := level
+	mu.RUnlock()
+	if lvl != LevelDebug {
+		return
+	}
+	printMsg("DEBUG", fmt, a...)
+}
+
+// Warn prints a warning message.
+func Warn(fmt string, a ...interface{}) {
+	printMsg("WARN", fmt, a...)
+}
+
+var (
+	errmu   sync.RWMutex                // guards below fields
+	erragg  = map[string]*errorReport{} // aggregated errors
+	errrate time.Duration               // the rate at which errors are reported
+	erron   bool                        // true if errors are being aggregated
+)
+
+func init() {
```

Comment:
I agree. I'm not a fan of it either. I also know it's [frowned upon](https://dave.cheney.net/2017/06/11/go-without-package-scoped-variables) in the Go community to have package-level variables. But I think that some aspects make it more acceptable because:

* Everything is unexported
* Easily testable
* Significantly simplifies the API